### PR TITLE
Encode calling convention in mangled function pointer representation

### LINKED ISSUE #81106: Test failure Loader\\classloader\\Casting\\Functionpointer\\Functionpointer.cmd
Failed in Run: [runtime-coreclr outerloop 20230124.1](https://dev.azure.com/dnceng-public/public/_build/results?buildId=146350&view=results) and many ohter outerloop runs

Failed tests:
```
R2R-CG2 linux x64 Checked no_tiered_compilation @ Ubuntu.1804.Amd64.Open
R2R-CG2 windows x86 Checked no_tiered_compilation @ Windows.10.Amd64.Open
R2R-CG2 linux x64 Checked no_tiered_compilation @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64
R2R-CG2 linux x64 Checked @ Ubuntu.1804.Amd64.Open
R2R-CG2 windows x86 Checked @ Windows.10.Amd64.Open
R2R-CG2 windows x64 Checked @ Windows.10.Amd64.Open
R2R-CG2 linux arm64 Checked no_tiered_compilation @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8
R2R-CG2 linux arm64 Checked @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8
R2R-CG2 windows arm Checked no_tiered_compilation @ Windows.11.Arm64.Open
R2R-CG2 linux x64 Checked @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64
R2R-CG2 linux x64 Checked @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64
R2R-CG2 windows x64 Checked no_tiered_compilation @ Windows.10.Amd64.Open
R2R-CG2 windows arm Checked @ Windows.11.Arm64.Open
R2R-CG2 windows arm64 Checked no_tiered_compilation @ Windows.11.Arm64.Open
R2R-CG2 linux arm64 Checked @ (Alpine.314.Arm64.Open)Ubuntu.1804.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-arm64v8
R2R-CG2 osx x64 Checked @ OSX.1200.Amd64.Open
R2R-CG2 osx x64 Checked no_tiered_compilation @ OSX.1200.Amd64.Open
R2R-CG2 windows arm64 Checked @ Windows.11.Arm64.Open
  - Loader\classloader\Casting\Functionpointer\Functionpointer.cmd [FAIL]
```

Error message:
```
      Process terminated. Assertion failed.
      Duplicate node name emitted to file
      Symbol DelayLoadHelperImport(DelayLoad_Helper) -> SignaturePointer_NewArraySignature: __Array<__FnPtr<Int32<>>> has already been written to the output object file C:\h\w\950D07E7\w\AF0A0977\e\Loader\classloader\Casting\Functionpointer\Functionpointer.dll with symbol ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter+NodeInfo
         at ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitObjectData(R2RPEBuilder r2rPeBuilder, ObjectData data, Int32 nodeIndex, String name, ObjectNodeSection section) in /_/src/coreclr/tools/aot/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:line 500
         at ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitPortableExecutable() in /_/src/coreclr/tools/aot/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:line 296
         at ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitObject(String objectFilePath, EcmaModule componentModule, IEnumerable`1 inputFiles, IEnumerable`1 nodes, NodeFactory factory, Boolean generateMapFile, Boolean generateMapCsvFile, Boolean generatePdbFile, String pdbPath, Boolean generatePerfMapFile, String perfMapPath, Int32 perfMapFormatVersion, Boolean generateProfileFile, CallChainProfile callChainProfile, Int32 customPESectionAlignment) in /_/src/coreclr/tools/aot/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:line 547
         at ILCompiler.ReadyToRunCodegenCompilation.Compile(String outputFile) in /_/src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:line 356
         at ILCompiler.Program.RunSingleCompilation(Dictionary`2 inFilePaths, InstructionSetSupport instructionSetSupport, String compositeRootPath, Dictionary`2 unrootedInputFilePaths, HashSet`1 versionBubbleModulesHash, ReadyToRunCompilerContext typeSystemContext) in /_/src/coreclr/tools/aot/crossgen2/Program.cs:line 616
         at ILCompiler.Program.Run() in /_/src/coreclr/tools/aot/crossgen2/Program.cs:line 289
         at ILCompiler.Crossgen2RootCommand.<>c__DisplayClass187_0.<.ctor>b__0(InvocationContext context) in /_/src/coreclr/tools/aot/crossgen2/Crossgen2RootCommand.cs:line 293
         at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext context)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<BuildInvocationChain>b__0(InvocationContext invocationContext, Func`2 _)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_2.<BuildInvocationChain>b__3(InvocationContext c)
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseErrorReporting>b__0>d.MoveNext()
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<UseParseErrorReporting>b__0(InvocationContext context, Func`2 next)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_2.<BuildInvocationChain>b__3(InvocationContext c)
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseHelp>b__0>d.MoveNext()
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<UseHelp>b__0(InvocationContext context, Func`2 next)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_2.<BuildInvocationChain>b__3(InvocationContext c)
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<UseVersionOption>b__0(InvocationContext context, Func`2 next)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_1.<BuildInvocationChain>b__2(InvocationContext ctx, Func`2 next)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_1.<BuildInvocationChain>b__2(InvocationContext ctx, Func`2 next)
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_1.<BuildInvocationChain>b__2(InvocationContext ctx, Func`2 next)
         at System.CommandLine.Invocation.InvocationPipeline.<Invoke>g__FullInvocationChain|3_0(InvocationContext context)
         at System.CommandLine.Invocation.InvocationPipeline.Invoke(IConsole console)
         at System.CommandLine.Parsing.ParseResultExtensions.Invoke(ParseResult parseResult, IConsole console)
         at System.CommandLine.Parsing.ParserExtensions.Invoke(Parser parser, String[] args, IConsole console)
         at ILCompiler.Program.Main(String[] args) in /_/src/coreclr/tools/aot/crossgen2/Program.cs:line 892
```

Stack trace:
```
           at Loader_classloader._Casting_Functionpointer_Functionpointer_._Casting_Functionpointer_Functionpointer_cmd()
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
           at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

### PR DESCRIPTION
**Pull request by trylek**
This fixes the recently introduced issue #81106 tracking Crossgen2 crash due to not being able to distinguish two function pointers differing only by calling convention (one of the pointers is Static and the other is Static | UnmanagedCallingConvention).

Thanks

Tomas

Fixes: #81106

/cc @dotnet/crossgen-contrib

### CODE REVIEW

File: src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs
Comment (fanyang-mono) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:295]: Are runtime tests running with NativeAOT? It seems that there is a gap in NativeAOT as well. I don't recall seeing the function pointer tests failing for NativeAOT.

Comment (trylek) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:295]: @fanyang-mono I defer to Michal to elaborate in more detail, according to the running times and the logs I'm seeing for the NativeAOT legs in the runtime (PR / CI) pipeline I'm guessing that right now just a small subset of Pri0 tests gets run on a regular basis in NativeAOT mode.

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:295]: This codepath is only exercised in obscure cases in NativeAOT - https://github.com/dotnet/runtime/pull/71749. Most of the time function pointers are treated as IntPtr right now. It will have to change for .NET 8.

And yes, we don't run regular Pri0/Pri1 CoreCLR test suite runs right now. Another thing getting addressed for 8.0.

Comment (fanyang-mono) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:295]: Gotcha!

**Comment by trylek** [comment]
@steveharter - could you please respond to @MichalStrehovsky's question when you have a chance? We're trying to stay on top of customer bug escalations and I'm having an active Teams thread with Mono folks who need the fix getting backported to .NET 7 so I'm trying to figure out how to make it at least to main.

File: src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs
Comment (jkotas) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:295]: > @steveharter what was the decision about function pointer type identity within the runtime?

@steveharter is on vacation.

In current main, the function pointer runtime type identity includes calling convention from signature. Modopts are not included in the function pointer runtime type identity.

After https://github.com/dotnet/runtime/pull/81006, the function pointer runtime type identity is only going to include managed vs. unmanaged. The full calling convention details and modopts are not going to be part of function pointer identity.

File: src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:293]: Given https://github.com/dotnet/runtime/pull/81122#discussion_r1090203792 I think this comment can be deleted.

I think we're going to require that after #81006 whoever is calling into this logic already normalized the function pointer type so that differences between `delegate* unmanaged[Cdecl]<void>` and `delegate* unmanaged<void>` don't exist.

**Review by MichalStrehovsky — APPROVED** [review]

File: src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs
Comment (trylek) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs:294]: Thanks Jan and Michal for your feedback. I have deleted the superfluous comment per Michal's suggestion in the 2nd commit.

File: src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs
Comment (trylek) [reply src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs]: Deleted in 2nd commit.

**Comment by trylek** [comment]
The one failure in the DictionaryExpansion test on Windows x86 is known,

https://github.com/dotnet/runtime/issues/75244

I'll probably take the bug as it's currently assigned to @davidwrighton who's expected to be OOF for a prolonged period of time. I'm going to merge this in assuming no other untracked issues pop up in the remaining outerloop jobs.

**Comment by trylek** [comment]
OK, there's one more bug but that's also known, #81109, I'm actually investigating it right now, that's obviously unrelated to this change, merging in.

**Pull request merged**
